// @fern-api/mrlint-rules/lib/utils/Executables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Executables = exports.EXECUTABLES = exports.Executable = void 0;
var Executable;
(function (Executable) {
    Executable[Executable["TSC"] = 0] = "TSC";
    Executable[Executable["JEST"] = 1] = "JEST";
    Executable[Executable["ESLINT"] = 2] = "ESLINT";
    Executable[Executable["STYLELINT"] = 3] = "STYLELINT";
    Executable[Executable["PRETTIER"] = 4] = "PRETTIER";
    Executable[Executable["DEPCHECK"] = 5] = "DEPCHECK";
    Executable[Executable["ENV_CMD"] = 6] = "ENV_CMD";
    Executable[Executable["AWS_CDK"] = 7] = "AWS_CDK";
    Executable[Executable["REACT_SCRIPTS"] = 8] = "REACT_SCRIPTS";
})(Executable = exports.Executable || (exports.Executable = {}));
exports.EXECUTABLES = {
    [Executable.TSC]: "tsc",
    [Executable.JEST]: "jest",
    [Executable.ESLINT]: "eslint",
    [Executable.STYLELINT]: "stylelint",
    [Executable.PRETTIER]: "prettier",
    [Executable.DEPCHECK]: "depcheck",
    [Executable.ENV_CMD]: "env-cmd",
    [Executable.AWS_CDK]: "cdk",
    [Executable.REACT_SCRIPTS]: "react-scripts",
};
const DEPENDENCIES = {
    [Executable.TSC]: "typescript",
    [Executable.JEST]: "jest",
    [Executable.ESLINT]: "eslint",
    [Executable.STYLELINT]: "stylelint",
    [Executable.PRETTIER]: "prettier",
    [Executable.DEPCHECK]: "depcheck",
    [Executable.ENV_CMD]: "env-cmd",
    [Executable.AWS_CDK]: "aws-cdk",
    [Executable.REACT_SCRIPTS]: "react-scripts",
};
class Executables {
    constructor() {
        this.accessed = new Set();
    }
    get(executable) {
        this.accessed.add(executable);
        return exports.EXECUTABLES[executable];
    }
    getRequiredDependencies() {
        return [...this.accessed].map((executable) => ({
            dependency: DEPENDENCIES[executable],
            executable,
        }));
    }
}
exports.Executables = Executables;
