// @fern-api/mrlint-rules/lib/rules/craco.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CracoRule = void 0;
const mrlint_commons_1 = require("@fern-api/mrlint-commons");
const writePackageFile_1 = require("../utils/writePackageFile");
exports.CracoRule = {
    ruleId: "craco",
    type: mrlint_commons_1.RuleType.PACKAGE,
    targetedPackages: [mrlint_commons_1.PackageType.REACT_APP],
    run: runRule,
};
const CONTENTS = `const { getLoader, loaderByName } = require("@craco/craco");
const NodePolyfillPlugin = require("node-polyfill-webpack-plugin");
const { getPackages } = require("@lerna/project");

module.exports = async function () {
    const packages = (await getPackages()).map((p) => \`\${p.location}/src\`);
    return {
        webpack: {
            plugins: [new NodePolyfillPlugin()],
            configure: (webpackConfig) => {
                const { isFound, match } = getLoader(webpackConfig, loaderByName("babel-loader"));
                if (isFound) {
                    const include = Array.isArray(match.loader.include) ? match.loader.include : [match.loader.include];
                    match.loader.include = include.concat(packages);
                } else {
                    throw new Error("Could not find babel-loader");
                }

                // add build flag to tsc
                const forkTsCheckerWebpackPlugin = webpackConfig.plugins.find(
                    (p) => p.constructor.name === "ForkTsCheckerWebpackPlugin"
                );
                if (forkTsCheckerWebpackPlugin != null) {
                    forkTsCheckerWebpackPlugin.options.typescript.build = true;
                } else {
                    throw new Error("Could not find ForkTsCheckerWebpackPlugin");
                }
                return webpackConfig;
            },
        },
    };
};`;
async function runRule({ fileSystems, packageToLint, logger }) {
    return (0, writePackageFile_1.writePackageFile)({
        fileSystem: fileSystems.getFileSystemForPackage(packageToLint),
        filename: "craco.config.js",
        contents: CONTENTS,
        logger,
    });
}
