// @fern-api/mrlint-rules/lib/rules/env-cmd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnvCmdRule = void 0;
const mrlint_commons_1 = require("@fern-api/mrlint-commons");
const writePackageFile_1 = require("../utils/writePackageFile");
const FILENAME = ".env-cmdrc";
exports.EnvCmdRule = {
    ruleId: "env-cmd",
    type: mrlint_commons_1.RuleType.PACKAGE,
    targetedPackages: [mrlint_commons_1.PackageType.REACT_APP],
    run: runRule,
};
async function runRule({ fileSystems, packageToLint, logger }) {
    var _a, _b;
    const fileSystemForPackage = fileSystems.getFileSystemForPackage(packageToLint);
    const existingStr = await fileSystemForPackage.readFile(FILENAME);
    const existing = existingStr != null ? JSON.parse(existingStr) : undefined;
    const contents = Object.assign(Object.assign({}, existing), { development: (_a = existing.development) !== null && _a !== void 0 ? _a : {}, production: (_b = existing.production) !== null && _b !== void 0 ? _b : {} });
    return (0, writePackageFile_1.writePackageFile)({
        fileSystem: fileSystems.getFileSystemForPackage(packageToLint),
        filename: FILENAME,
        contents: JSON.stringify(contents),
        logger,
    });
}
