// @fern-api/mrlint-rules/lib/rules/prettier.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrettierRule = void 0;
const mrlint_commons_1 = require("@fern-api/mrlint-commons");
const path_1 = __importDefault(require("path"));
const writePackageFile_1 = require("../utils/writePackageFile");
exports.PrettierRule = {
    ruleId: "prettier",
    type: mrlint_commons_1.RuleType.PACKAGE,
    targetedPackages: [
        mrlint_commons_1.PackageType.REACT_APP,
        mrlint_commons_1.PackageType.REACT_LIBRARY,
        mrlint_commons_1.PackageType.TYPESCRIPT_LIBRARY,
        mrlint_commons_1.PackageType.TYPESCRIPT_CLI,
    ],
    run: runRule,
};
async function runRule({ fileSystems, packageToLint, relativePathToRoot, logger, }) {
    const contents = `module.exports = {
...require("${path_1.default.join(relativePathToRoot, ".prettierrc.json")}"),
};`;
    return (0, writePackageFile_1.writePackageFile)({
        fileSystem: fileSystems.getFileSystemForPackage(packageToLint),
        filename: ".prettierrc.js",
        contents,
        logger,
    });
}
