// @fern-api/mrlint-rules/lib/rules/stylelint.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StyleLintRule = void 0;
const mrlint_commons_1 = require("@fern-api/mrlint-commons");
const path_1 = __importDefault(require("path"));
const writePackageFile_1 = require("../utils/writePackageFile");
exports.StyleLintRule = {
    ruleId: "stylelint",
    type: mrlint_commons_1.RuleType.PACKAGE,
    targetedPackages: [mrlint_commons_1.PackageType.REACT_APP, mrlint_commons_1.PackageType.REACT_LIBRARY],
    run: runRule,
};
async function runRule({ fileSystems, packageToLint, relativePathToSharedConfigs, logger, }) {
    const contents = {
        extends: [path_1.default.join(relativePathToSharedConfigs, "stylelintrc.shared.json")],
    };
    return (0, writePackageFile_1.writePackageFile)({
        fileSystem: fileSystems.getFileSystemForPackage(packageToLint),
        filename: ".stylelintrc.json",
        contents: JSON.stringify(contents),
        logger,
    });
}
